internal/bkmultipart/streamer.go

```diff
@@ -0,0 +1,92 @@
+package bkmultipart
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"os"
+)
+
+// A wrapper around the complexities of streaming a multipart file and fields to an http endpoint that infuriatingly
+// requires a Content-Length
+// Stolen/adapted from the multipart streamer in https://github.com/buildkite/agentv/v3/internal/artifact, which itself
+// was derived from https://github.com/technoweenie/multipartstreamer
+type Streamer struct {
+	ContentType string
+
+	bodyBuffer  *bytes.Buffer
+	closeBuffer *bytes.Buffer
+	bodyWriter  *multipart.Writer
+
+	reader        io.Reader
+	contentLength int64
+	writtenFile   bool
+}
+
+// NewStreamer initializes a new Streamer.
+func NewStreamer() *Streamer {
+	s := &Streamer{bodyBuffer: &bytes.Buffer{}}
+
+	s.bodyWriter = multipart.NewWriter(s.bodyBuffer)
+	boundary := s.bodyWriter.Boundary()
+	s.ContentType = "multipart/form-data; boundary=" + boundary
+
+	closeBoundary := fmt.Sprintf("\r\n--%s--\r\n", boundary)
+	s.closeBuffer = bytes.NewBufferString(closeBoundary)
+
+	// if a file is added, the reader will be set up to read from the body, the file and the closer
```

Comment:
yeah... i might refactor this so that it has an internal slice of `io.Reader`s that it appends to, and have it have a method like 
```Go
func (s *Streamer) Reader io.Reader {
  r := io.MultiReader(s.readers...)
  r = io.MultiReader(r, closeBuffer)
}
```

which would remove the necessity to only call `WriteFile` once, and allow the files to be put anywhere in the body (though in our case they always need to be at the end)